src/containers/NonBitGoRecoveryCoin/SolanaForm.tsx

```diff
@@ -0,0 +1,191 @@
+import { Form, FormikHelpers, FormikProvider, useFormik } from 'formik';
+import { Link } from 'react-router-dom';
+import * as Yup from 'yup';
+import {
+  Button,
+  FormikSelectfield,
+  FormikTextarea,
+  FormikTextfield,
+  Icon,
+  Notice,
+} from '~/components';
+
+const validationSchema = Yup.object({
+  backupKey: Yup.string().required(),
+  bitgoKey: Yup.string().required(),
+  krsProvider: Yup.string()
+    .oneOf(['keyternal', 'bitgoKRSv2', 'dai'])
+    .label('Key Recovery Service'),
+  publicKey: Yup.string().required(),
+  recoveryDestination: Yup.string().required(),
+  scan: Yup.number().required(),
+  userKey: Yup.string().required(),
+  walletPassphrase: Yup.string().required(),
+})
+  .required()
+  .shape(
+    {
+      publicKey: Yup.string().when('secretKey', {
+        is: (val: string) => !!val,
+        then: Yup.string().required(
+          'A public key must be provided with a secret key'
+        ),
+      }),
+      secretKey: Yup.string().when('publicKey', {
+        is: (val: string) => !!val,
+        then: Yup.string().required(
+          'A secret key must be provided with a public key'
+        ),
+      }),
+    },
+    [['secretKey', 'publicKey']]
+  );
```

Comment:
Yeah, basically if one of the fields is defined the other must be as well. By using when(), I make one field a dependency of the other. This introduces a cyclical dependency between publicKey and secretKey. To get around this, shape() provides the optional parameter: noSortEdges?: Array<[string, string]>. The following explanation is from https://github.com/jquense/yup/issues/720:

It's a ugly escape hatch for dealing with cyclical dependencies. Generally fields are topographically sorted so they can be resolved in the correct order, i.e. field A resolves before field B if field B depends on it.

In some cases tho the order doesn't matter, because the dependency is purely for the sake of validation, which all happen after all casting is done. In those cases it's fine if there are cycles between fields, and you can manually opt out of the ordering, if you are sure that it's not needed. The reason it's not documented is that its a bad API and really easy to shoot one's self in the foot with it. I've had dreams of adding a real API but it's a bit tricky b/c it would require determining if a dependency is validation only, at schema construction, something that isn't really possible.

Reply:
okay, but then why is publicKey required in the Yup.object({}) if both are potentially optional?

Reply:
Oh you're correct, that's an error.